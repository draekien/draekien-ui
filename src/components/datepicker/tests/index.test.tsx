/** @jsx jsx */
import { jsx } from 'theme-ui';
import * as React from 'react';
import DatePicker from '..';
import {
  fireEvent,
  getAllByText,
  getByText,
  render,
} from '@testing-library/react';

describe('DatePicker component', () => {
  test('should match snapshot', () => {
    const { container } = render(<DatePicker />);

    expect(container).toMatchSnapshot();
  });

  test('should render the correct month when one is selected', () => {
    const { container } = render(
      <DatePicker selectedDates={[new Date('22 Nov 2020')]} />
    );

    expect(container.textContent).toContain('November 2020');
  });

  test('should render correct month when multiple selected dates cross multiple months', () => {
    const { container } = render(
      <DatePicker
        selectedDates={[
          new Date('22 Nov 2019'),
          new Date('11 Nov 2019'),
          new Date('2 Dec 2019'),
          new Date('3 Dec 2019'),
        ]}
      />
    );

    expect(container.textContent).toContain('December 2019');
  });

  test('should show next month when next month btn pressed', (done) => {
    const fn = jest.fn(() => done());
    const component = render(
      <DatePicker
        selectedDates={[new Date('22 Nov 2019')]}
        onChangeMonth={fn}
      />
    );

    fireEvent(
      getByText(component.container as HTMLElement, 'chevron_right'),
      new MouseEvent('click', { bubbles: true, cancelable: true })
    );

    expect(component.container.textContent).toContain('December 2019');
  });

  test('should show prev month when prev month btn pressed', (done) => {
    const fn = jest.fn(() => done());
    const component = render(
      <DatePicker
        selectedDates={[new Date('22 Nov 2019')]}
        onChangeMonth={fn}
      />
    );

    fireEvent(
      getByText(component.container as HTMLElement, 'chevron_left'),
      new MouseEvent('click', { bubbles: true, cancelable: true })
    );

    expect(component.container.textContent).toContain('October 2019');
  });

  test('should render correct days for month when Sunday is first day of week', () => {
    const component = render(
      <DatePicker selectedDates={[new Date('22 Oct 2019')]} />
    );

    expect(component.container.textContent).toContain(
      '29301234567891011121314151617181920212223242526272829303112'
    );
  });

  test('should render correct days for month when Monday is first day of week', () => {
    const component = render(
      <DatePicker
        weekStartDay="Monday"
        selectedDates={[new Date('22 Oct 2019')]}
      />
    );

    expect(component.container.textContent).toContain(
      '3012345678910111213141516171819202122232425262728293031123'
    );
  });

  test('should go to prev month when date from prev month is pressed', () => {
    const component = render(
      <DatePicker selectedDates={[new Date('22 Nov 2019')]} />
    );

    fireEvent(
      getAllByText(component.container as HTMLElement, '27')[0],
      new MouseEvent('click', { bubbles: true, cancelable: true })
    );

    expect(component.container.textContent).toContain('October 2019');
  });

  test('calls callback with correct date in single mode', () => {
    const fn = jest.fn();
    const date = new Date('2 Dec 2019');
    const component = render(
      <DatePicker selectedDates={[new Date('3 Dec 2019')]} onChange={fn} />
    );

    fireEvent(
      getAllByText(component.container as HTMLElement, '2')[0],
      new MouseEvent('click', { bubbles: true, cancelable: true })
    );

    expect(fn).toHaveBeenCalledTimes(1);
    expect(fn).toHaveBeenCalledWith([date], date);
  });
});
