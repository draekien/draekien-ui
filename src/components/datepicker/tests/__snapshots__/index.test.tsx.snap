// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatePicker component should match snapshot 1`] = `
<div>
  <div
    class="css-1ffeio2-DatePicker"
  >
    <div
      class="css-1irkkch-DatePicker"
    >
      <button
        class="css-r4utam-Button"
        type="button"
      >
        <span
          class="css-1ar16h4"
        >
          <i
            class="material-icons  css-1ike9wb-Icon"
          >
            chevron_left
          </i>
        </span>
        <span />
      </button>
      <span
        class="css-eydsp4-Text"
      >
        December 2020
      </span>
      <button
        class="css-r4utam-Button"
        type="button"
      >
        <span />
        <span
          class="css-1ar16h4"
        >
          <i
            class="material-icons  css-1ike9wb-Icon"
          >
            chevron_right
          </i>
        </span>
      </button>
    </div>
    <div
      class="css-o64vi5-DatePicker"
    >
      <span
        class="css-eydsp4-Text"
      >
        S
      </span>
      <span
        class="css-eydsp4-Text"
      >
        M
      </span>
      <span
        class="css-eydsp4-Text"
      >
        T
      </span>
      <span
        class="css-eydsp4-Text"
      >
        W
      </span>
      <span
        class="css-eydsp4-Text"
      >
        T
      </span>
      <span
        class="css-eydsp4-Text"
      >
        F
      </span>
      <span
        class="css-eydsp4-Text"
      >
        S
      </span>
    </div>
    <div
      class="css-1t85os0-DatePicker"
    >
      <span
        class="css-7r7jbe-Day"
      >
        29
      </span>
      <span
        class="css-7r7jbe-Day"
      >
        30
      </span>
      <span
        class="css-2hjx6u-Day"
      >
        1
      </span>
      <span
        class="css-2hjx6u-Day"
      >
        2
      </span>
      <span
        class="css-2hjx6u-Day"
      >
        3
      </span>
      <span
        class="css-2hjx6u-Day"
      >
        4
      </span>
      <span
        class="css-2hjx6u-Day"
      >
        5
      </span>
      <span
        class="css-2hjx6u-Day"
      >
        6
      </span>
      <span
        class="css-2hjx6u-Day"
      >
        7
      </span>
      <span
        class="css-2hjx6u-Day"
      >
        8
      </span>
      <span
        class="css-2hjx6u-Day"
      >
        9
      </span>
      <span
        class="css-2hjx6u-Day"
      >
        10
      </span>
      <span
        class="css-2hjx6u-Day"
      >
        11
      </span>
      <span
        class="css-2hjx6u-Day"
      >
        12
      </span>
      <span
        class="css-2hjx6u-Day"
      >
        13
      </span>
      <span
        class="css-2hjx6u-Day"
      >
        14
      </span>
      <span
        class="css-2hjx6u-Day"
      >
        15
      </span>
      <span
        class="css-2hjx6u-Day"
      >
        16
      </span>
      <span
        class="css-2hjx6u-Day"
      >
        17
      </span>
      <span
        class="css-2hjx6u-Day"
      >
        18
      </span>
      <span
        class="css-2hjx6u-Day"
      >
        19
      </span>
      <span
        class="css-2hjx6u-Day"
      >
        20
      </span>
      <span
        class="css-2hjx6u-Day"
      >
        21
      </span>
      <span
        class="css-2hjx6u-Day"
      >
        22
      </span>
      <span
        class="css-2hjx6u-Day"
      >
        23
      </span>
      <span
        class="css-2hjx6u-Day"
      >
        24
      </span>
      <span
        class="css-2hjx6u-Day"
      >
        25
      </span>
      <span
        class="css-2hjx6u-Day"
      >
        26
      </span>
      <span
        class="css-2hjx6u-Day"
      >
        27
      </span>
      <span
        class="css-2hjx6u-Day"
      >
        28
      </span>
      <span
        class="css-2hjx6u-Day"
      >
        29
      </span>
      <span
        class="css-2hjx6u-Day"
      >
        30
      </span>
      <span
        class="css-2hjx6u-Day"
      >
        31
      </span>
      <span
        class="css-7r7jbe-Day"
      >
        1
      </span>
      <span
        class="css-7r7jbe-Day"
      >
        2
      </span>
    </div>
  </div>
</div>
`;
